dvc/repo/metrics/show.py

```diff
@@ -23,16 +23,23 @@ def _collect_metrics(repo, targets, recursive):
     if not targets:
         return list(metrics)
 
+    result = []
     target_infos = [PathInfo(os.path.abspath(target)) for target in targets]
+    if recursive:
 
-    def _filter(path_info):
-        for info in target_infos:
-            func = path_info.isin_or_eq if recursive else path_info.__eq__
-            if func(info):
-                return True
-        return False
+        def _filter(path_info):
+            for info in target_infos:
+                if path_info.isin_or_eq(info):
+                    return True
+            return False
 
-    return list(filter(_filter, metrics))
+        result.extend(list(filter(_filter, metrics)))
+
+    tree = RepoTree(repo)
+    for target_info in target_infos:
+        if tree.isfile(target_info):
```

Comment:
What if a target is a file and `recursive==True`? Will it be added to the `result` twice?

Reply:
This will probably become obsoleted once you implement recursive for unofficial metrics above...